Add username validation feature

Smnrgcl opened the pull request:
This commit introduces a new validation rule in the service layer to ensure that usernames do not contain any special characters, enhancing the system's security and data integrity. It also includes comprehensive unit tests for the new validation logic, covering various edge cases to ensure robustness and reliability.

ebrarkesici committed:
Fixes #8: I fixed the issue.

ebrarkesici committed:
Closes #8. The dependencies in the Makefile were resolved.

Sddilora commented:
Enhancements are needed in error handling as the message 'Username is not valid' lacks specificity, potentially confusing users about the rules and necessary corrections.

Sddilora commented:
There are the definitions of the Error variables in, [domain package](https://github.com/octoposprime/op-be-user/blob/phase1/internal/domain/model/object/error.go). 

```go
        ErrorUserUsernameIsEmpty                            error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrEmpty)
	ErrorUserUsernameIsExists                           error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrAlreadyExists)
	ErrorUserUsernameIsTooShort                         error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooShort)
	ErrorUserUsernameIsTooLong                          error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooLong)
	ErrorUserUsernameIsNotValid                         error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrNotValid)
```
"Username cannot contain special characters" needs to be implemented in this way too. You can find the smdol.Errbase or other declarations in [model package of the shared library](https://github.com/octoposprime/op-be-shared/blob/phase1/pkg/model/error.go)

Sddilora reviewed (changes requested):
Error handling methods should adhere to the project structure and be enhanced. 
❗❗❗Please add the IDs of the issue that will be closed and resolved with this pull request. ( To close the issue testFunctions for those functions need to be implemented too #5 )

husamettinarabaci reviewed (changes requested):
We have struct-based error types. Please keep our generic mo.Error*** styles. You have to add all new errors in model/object/error.go.

Note: in the Golang; any error message shouldn't start with a capital character.

Smnrgcl committed:
CheckUserNameRules enhancet, test func added

desired changes maden in CheckUserNameRules' error handling mechanism and  test func added

husamettinarabaci commented on internal/domain/service/service_test.go:
```diff
@@ -1,103 +1,73 @@
-package domain_test
+ppackage domain
```
Please check this

husamettinarabaci reviewed (changes requested):
Please caution this message;
https://github.com/octoposprime/op-be-user/pull/16#issuecomment-1975302539

Smnrgcl committed:
corrected 'ppackage' to 'package domain_test' in the codebase.

husamettinarabaci commented:
Please remove;

var (
    ErrorUserUsernameIsEmpty            = errors.New("username is empty")
    ErrorUserUsernameIsTooShort         = errors.New("username is too short")
    ErrorUserUsernameIsTooLong          = errors.New("username is too long")
    ErrorUserUsernameContainsSpace      = errors.New("username contains space")
    ErrorUserUsernameContainsSpecialChars = errors.New("username contains special characters")
    ErrorUserUsernameContainsInvalidChars = errors.New("username contains invalid characters")
)

Smnrgcl committed:
Merge branch 'phase1' of https://github.com/octoposprime/op-be-user into feature/smnr/5/username-validation

Smnrgcl committed:
Test file fixed

Smnrgcl committed:
error go file fixed

Sddilora commented:
![image](https://github.com/octoposprime/op-be-user/assets/122971619/df1f0c0f-1a5b-4408-af45-ac88bab1465c)
As in Godev's example, the name of the test suite should be the same as the name of the source package (the name of the package containing the function under test).

Also, such things should not be changed :)

Sddilora commented on internal/domain/service/service_test.go:
```diff
@@ -1,29 +1,34 @@
-package domain
```
This part should be remained as "domain".

Sddilora commented on internal/domain/service/service_test.go:
```diff
@@ -1,29 +1,34 @@
-package domain
+package domain_test
```
This part should be remained as "domain".

Sddilora commented on internal/domain/model/object/error.go:
```diff
@@ -67,6 +71,8 @@ var (
 	ErrorUserUsernameIsTooShort                         error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooShort)
 	ErrorUserUsernameIsTooLong                          error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooLong)
+	ErrorUserUsernameContainsInValidChars               error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrContainsInvalidChars)
+	ErrorUserUsernameContainsSpecialChars               error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrorUserUsernameContainsSpecialChars )         
```
initialization cycle: ErrorUserUsernameContainsSpecialChars refers to itselfcompiler[InvalidInitCycle](https://pkg.go.dev/golang.org/x/tools/internal/typesinternal#InvalidInitCycle)

Sddilora commented on internal/domain/service/service.go:
```diff
@@ -1,93 +1,140 @@
+    }
+    // Check for spaces
+    if strings.Contains(user.UserName, " ") {
+        return mo.ErrorUserUsernameContainsSpace
```
this is not present in mo package

Sddilora commented on internal/domain/model/object/error.go:
```diff
@@ -67,6 +71,8 @@ var (
 	ErrorUserUsernameIsExists                           error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrAlreadyExists)
 	ErrorUserUsernameIsTooShort                         error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooShort)
 	ErrorUserUsernameIsTooLong                          error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrTooLong)
+	ErrorUserUsernameContainsInValidChars               error = errors.New(smodel.ErrBase + smodel.ErrSep + ErrUser + smodel.ErrSep + ErrUserName + smodel.ErrSep + ErrContainsInvalidChars)
```
typo causes error

Smnrgcl committed:
Added POSTGRES_DATABASE field, removed PostgreSQL fields

Smnrgcl committed:
Fixed typos and errors in the code

Sddilora commented:
Let's start fresh with a new branch and close this pull request, as the current branch and PR are causing some confusion.
